src/levanter/eval_harness.py

```diff
@@ -0,0 +1,564 @@
+                model_config.model_type, ref=config.checkpoint_path, dtype=config.trainer.mp.compute_dtype  # type: ignore
+            )
+        else:
+            with use_cpu_device():
+                model = eqx.filter_eval_shape(config.model.build, Vocab, key=key)
+                model = load_checkpoint(model, config.checkpoint_path, subpath="model")
+            model = hax.shard(model, parameter_axis_mapping)
+
+        model = typing.cast(LmHeadModel, inference_mode(model, True))
+
+        logger.info("Running LM eval harness....")
+        outputs = run_lm_eval_harness(
+            config.eval_harness,
+            model,
+            tokenizer,
+            config.EvalBatch,
+            axis_resources=compute_axis_mapping,
+        )
+
+        logger.info("Finished running LM eval harness")
+        # log the results as json
+        with open("lm_eval_results.json", "w") as f:
+            json.dump(outputs, f, indent=2)
+
+        # also write to stdout
+        if jax.process_index() == 0:
+            print(json.dumps(outputs, indent=2), flush=True)
+
+        # also log the results
+        levanter.tracker.current_tracker().log_artifact("lm_eval_results.json")
+        log_report_to_tracker("lm_eval", outputs, levanter.tracker.current_tracker())
+
+        return outputs
+
+
+def log_report_to_tracker(prefix: str, report: dict, tracker: Optional[levanter.tracker.Tracker] = None):
+    if tracker is None:
+        tracker = levanter.tracker.current_tracker()
+
+    to_log = {}
+    for task_name, task_results in report["results"].items():
+        for metric_name, metric_value in task_results.items():
+            if metric_name.endswith(",none"):
+                metric_name = metric_name[:-5]
+
+            if isinstance(metric_value, float | int):
+                to_log[f"{prefix}/{task_name}/{metric_name}"] = metric_value
+
+    if "averages" in report:
+        for metric_name, metric_value in report["averages"].items():
+            if isinstance(metric_value, float | int):
+                if metric_name.endswith(",none"):
+                    metric_name = metric_name[:-5]
+
+                to_log[f"{prefix}/averages/{metric_name}"] = metric_value
+
+    tracker.log(to_log, step=None)
+
+
+def lm_eval_harness(config: LmEvalHarnessConfig, tokenizer, EvalBatch, axis_resources):
+    tasks_to_run = config.to_task_dict()
+
+    def lm_eval_harness(step: StepInfo, force=False):
+        # if step.step == 0 and not force:
+        #     return  # don't run eval on the first step
+
+        print(config.task_spec_or_default())
+
+        model = inference_mode(step.model, True)
+        outputs = _actually_run_eval_harness(
+            config,
+            model,
+            tasks_to_run,
+            tokenizer,
+            EvalBatch,
+            axis_resources,
+        )
+
+        if jax.process_index() == 0:
+            with tempfile.NamedTemporaryFile("w", delete=False, suffix=".json") as f:
```

Comment:
tmp file that doesn't get deleted doesn't sound good - I'd delete or put the file in the evaluation directory (which I guess would have to be passed in as an explicit location)

Reply:
it's delete=False so that the wandb process still can upload it